### PR TITLE
Increase multibyte_split performance by improving copy semantics.

### PR DESCRIPTION
**Pull request by cwharris**
reduces unnecessary copying by moving copy logic in to the `data_chunk_reader` API, introducing a new method `copy_next_chunk`, which on a per-implementation-basis can decide the best way to copy the underlying data source in to the required specific destination. This API is design to allow single-copy operations where using `get_next_chunk` may otherwise perform a copy under the hood before the data reaches it's final destination. Consumers should prefer `get_next_chunk` when the location of the data is not important, allowing device reads to take place in a zero-copy manner.

Thanks @ttnghia for the reminder. :)

before:
before
after:
after

### CODE REVIEW

File: cpp/include/cudf/io/text/data_chunk_source.hpp [hunk 1]
```diff
@@ -92,6 +92,21 @@ class data_chunk_reader {
    */
   virtual std::unique_ptr<device_data_chunk> get_next_chunk(std::size_t size,
                                                             rmm::cuda_stream_view stream) = 0;
+
+  /**
+   * @pure @brief Copy the next chunk of bytes from the data source in to a specific destination
```
Comment (ttnghia) [inline cpp/include/cudf/io/text/data_chunk_source.hpp:97]: What is `@pure`?

Comment (cwharris) [reply cpp/include/cudf/io/text/data_chunk_source.hpp:97]: No idea! It was in the other comment, so when I copied to modify, I left it in. @bdice any idea what this is?

Comment (bdice) [reply cpp/include/cudf/io/text/data_chunk_source.hpp:97]: https://www.doxygen.nl/manual/commands.html#cmdpure
> Indicates that the member documented by the comment block is pure virtual, i.e., it is abstract and has no implementation associated with it. This command is intended for use only when the language does not support the concept of pure virtual methods natively (e.g. C, PHP 4).

Seems like this doesn't apply here.
```suggestion
   * @brief Copy the next chunk of bytes from the data source in to a specific destination
```

Comment (harrism) [reply cpp/include/cudf/io/text/data_chunk_source.hpp:97]: > This command is intended for use only when the language does not support the concept of pure virtual methods natively (e.g. C, PHP 4).

Seems unnecessary.

**Review by harrism — APPROVED** [review]
Nice improvement.

**Review by ttnghia — APPROVED** [review]

**Comment by ttnghia** [comment]
Rerun tests.

**Review by upsj — APPROVED** [review]
LGTM! I suspect this might be superseded by future changes to the implementation, but for now it's a nice improvement.

**Comment by upsj** [comment]
@cwharris as this no longer applies, do you think the semantics would be useful elsewhere?

**Comment by cwharris** [comment]
@upsj yeah, this is something I wanted to add to the data chunk reader from the beginning, I just didn't have a use case for it. Anywhere we need to copy bytes in to a specific location, this is the right API. Otherwise we could be performing an extra copy. That said it has no place in the recently merged `multibyte_split` implementation.

**Pull request closed without merging**